src/TgSharp.Core/Auth/Authenticator.cs

```diff
@@ -1,45 +1,61 @@
+using System;
 using System.Threading;
 using System.Threading.Tasks;
 using TgSharp.Core.Network;
 
 namespace TgSharp.Core.Auth
 {
-    public static class Authenticator
+    public class Authenticator
     {
-        public static async Task<Step3_Response> DoAuthentication(TcpTransport transport, CancellationToken token = default(CancellationToken))
+        private readonly MtProtoPlainSender sender;
+        private TaskCompletionSource<Step3_Response> completionSource;
+        public Authenticator(WSTransport transport)
         {
-            token.ThrowIfCancellationRequested();
+            sender = new MtProtoPlainSender(transport);
+            completionSource = new TaskCompletionSource<Step3_Response>();
+        }
 
-            var sender = new MtProtoPlainSender(transport);
+        public Task<Step3_Response> DoAuthentication()
```

Comment:
if this returns Task, shouldn't it be marked as async?

Reply:
no you don't have to when you already have a Task, making the function async and awaiting a Task just adds more overhead.

Reply:
what do you mean more overhead? the way this method is consumed is actually via an await so it should be marked as async

Reply:
https://stackoverflow.com/questions/38017016/async-task-then-await-task-vs-task-then-return-task